share/shwap/p2p/bitswap/getter.go

```diff
@@ -0,0 +1,214 @@
+func (g *Getter) Start() {
+	ctx, cancel := context.WithCancel(context.Background())
+	g.availableSession = g.exchange.NewSession(ctx)
+	g.archivalSession = g.exchange.NewSession(ctx)
+	g.cancel = cancel
+}
+
+// Stop shuts down Getter's internal fetching session.
+func (g *Getter) Stop() {
+	g.cancel()
+}
+
+// GetShares uses [SampleBlock] and [Fetch] to get and verify samples for given coordinates.
+// TODO(@Wondertan): Rework API to get coordinates as a single param to make it ergonomic.
+func (g *Getter) GetShares(
+	ctx context.Context,
+	hdr *header.ExtendedHeader,
+	rowIdxs, colIdxs []int,
+) ([]share.Share, error) {
+	if len(rowIdxs) != len(colIdxs) {
+		return nil, fmt.Errorf("row indecies and col indecies must be same length")
+	}
+
+	if len(rowIdxs) == 0 {
+		return nil, fmt.Errorf("empty coordinates")
+	}
+
+	blks := make([]Block, len(rowIdxs))
+	for i, rowIdx := range rowIdxs {
+		sid, err := NewEmptySampleBlock(hdr.Height(), rowIdx, colIdxs[i], len(hdr.DAH.RowRoots))
+		if err != nil {
+			return nil, err
+		}
+
+		blks[i] = sid
+	}
+
+	ses := g.session(hdr)
+	err := Fetch(ctx, g.exchange, hdr.DAH, blks, WithStore(g.bstore), WithFetcher(ses))
+	if err != nil {
+		return nil, err
+	}
+
+	shares := make([]share.Share, len(blks))
+	for i, blk := range blks {
+		shares[i] = blk.(*SampleBlock).Container.Share
+	}
+
+	return shares, nil
+}
+
+// GetShare uses [GetShare] to fetch and verify single share by the given coordinates.
+func (g *Getter) GetShare(
+	ctx context.Context,
+	hdr *header.ExtendedHeader,
+	row, col int,
+) (share.Share, error) {
+	shrs, err := g.GetShares(ctx, hdr, []int{row}, []int{col})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(shrs) != 1 {
+		return nil, fmt.Errorf("expected 1 share row, got %d", len(shrs))
+	}
+
+	return shrs[0], nil
+}
+
+// GetEDS uses [RowBlock] and [Fetch] to get half of the first EDS quadrant(ODS) and
+// recomputes the whole EDS from it.
+// We fetch the ODS or Q1 to ensure better compatibility with archival nodes that only
+// store ODS and do not recompute other quadrants.
+func (g *Getter) GetEDS(
+	ctx context.Context,
+	hdr *header.ExtendedHeader,
+) (*rsmt2d.ExtendedDataSquare, error) {
+	sqrLn := len(hdr.DAH.RowRoots)
+	blks := make([]Block, sqrLn/2)
+	for i := 0; i < sqrLn/2; i++ {
+		blk, err := NewEmptyRowBlock(hdr.Height(), i, sqrLn)
+		if err != nil {
+			return nil, err
+		}
+
+		blks[i] = blk
+	}
+
+	ses := g.session(hdr)
+	err := Fetch(ctx, g.exchange, hdr.DAH, blks, WithFetcher(ses))
+	if err != nil {
+		return nil, err
+	}
+
+	shrs := make([]share.Share, 0, sqrLn/2*sqrLn/2)
+	for i, row := range blks {
+		rowShrs, err := row.(*RowBlock).Container.Shares()
+		if err != nil {
+			return nil, fmt.Errorf("decoding Shares out of Row: %w", err)
+		}
+
+		for j, shr := range rowShrs {
+			shrs[i*j] = shr
+		}
+	}
+
+	square, err := rsmt2d.ImportExtendedDataSquare(
+		shrs,
+		share.DefaultRSMT2DCodec(),
+		wrapper.NewConstructor(uint64(sqrLn/2)),
+	)
```

Comment:
Import would need all EDS shares, and you only requested half of rows. Consider using `ComputeExtendedDataSquare`

Reply:
Import works and that's tested. Compute expects only ods, while import can take empty shares